### ISSUE TITLE
Internal-definition context hygiene bug

### ISSUE THREAD

**Issue by michaelballantyne** [racket/racket#3198]
This program:

```
#lang racket
(require racket/block)

(define x 'good)
(define-syntax-rule (m) (displayln x))

(block
 (define x 'bad1)
 (m))

(define c%
  (class object%
    (super-new)
    (define x 'bad2)
    (m)))
(new c%)
```

Displays `'bad1` and `bad2`. I suspect other definition context macros behave similarly.

The underlying problem is that first-class definition contexts add inside-edge scopes but not outside-edge scopes.

We might think to fix this by adding an outside edge scope via `make-syntax-introducer`. Unfortunately that solution interacts badly with the way `'quote-syntax` removes scopes; the change breaks this test from `racket-test-core/tests/racket/macro.rktl`:

```
(module uses-internal-definition-context-around-id racket/base
  (require (for-syntax racket/base
                       racket/block))

  (define-syntax (m stx)
    (let ([x1 #'x]
          [x2 (block
               (define x3 #'x)
               x3)])
      #`(let ([#,x2 1]) #,x1)))

  (define v (m))
  (provide v))
(test 1 dynamic-require ''uses-internal-definition-context-around-id 'v)
```

The test breaks because the scope added via syntax introducer is not added to the expand context's set for `quote-syntax` to remove.

**Comment by mflatt** [comment]
When you say that an outside-edge scope interacts badly with `quote-syntax`, do you mean when `make-syntax-introducer` is used explicitly by `block`?

If so, is the solution to have `make-internal-definition-context` create an outside-edge scope that it records in the same way as the inside-edge scope, and then provide a function for `block` (and similar) to add that scope like the result of `make-syntax-introducer`?

Either way, it sounds like `block` and similar macros have to change to explicitly use an outside-edge scope. Is that right?

**Comment by michaelballantyne** [comment]
> When you say that an outside-edge scope interacts badly with quote-syntax, do you mean when make-syntax-introducer is used explicitly by block?

Yes.

> If so, is the solution to have make-internal-definition-context create an outside-edge scope that it records in the same way as the inside-edge scope, and then provide a function for block (and similar) to add that scope like the result of make-syntax-introducer?

This sounds like a reasonable solution.

> Either way, it sounds like block and similar macros have to change to explicitly use an outside-edge scope. Is that right?

Yes, I don't see a way to attach the scope to the correct syntax automatically. The macros will need to do it. Macros that currently use `make-syntax-introducer` for outside-edge scopes, like `splicing-let`, should also change to use the new function you propose.

I'm not sure what a correct use of `make-syntax-introducer` looks like, now. Uses that simulate outside-edge scopes are incorrect due to `quote-syntax` scope removal. Uses that simulate macro introduction scopes are incorrect because they don't come with corresponding use-site scopes.

**Comment by mflatt** [comment]
Since `syntax-local-make-definition-context` can't solve the problem automatically, I don't see a reason to tie the new scope to it, after all.

How about `syntax-local-make-introducer`, instead, which would be like `make-syntax-introducer`, but adds the scope to the set that is removed by `quote-syntax`?

**Comment by michaelballantyne** [comment]
I think either design works, and the choice between them depends on how we want users to understand the API; either:

  1. Thinking in terms of tasks, such as making a new scope / binding context / definition context. From this view connecting the outside-edge to the definition context makes sense, because it communicates that  when implementing a scope programmers should always use an outside-edge, inside-edge, and environment extension together.
  2. Thinking in terms of primitives such as outside-edge scopes, inside-edge scopes, environment extensions, etc. without dictating how they should be combined.

There are some other aspects of the definition context API that don't quite fit (1) right now, but it might be nice if it worked that way. (For example, parent context inside-edge scopes aren't added at local-expand; this means I can't simply use a chain of definition contexts to implement `define*` behavior without manually using internal-definition-context-introduce.)

**Comment by michaelballantyne** [comment]
> parent context inside-edge scopes aren't added at local-expand

Maybe this behavior could change at the same time? I've been thinking about it and I can't figure out the rationale for leaving out the parent context scopes, other than as an optimization. Could it have been necessary somehow because of the missing outside-edge scopes?

**Comment by mflatt** [comment]
Have you done any experiments to recommend either adding just a new primitive (which leaves it up to individual macros to be updated to use the new primitive) versus changing the existing behavior?

At this point, too many other Racket things have paged out the details of definition contexts for me, so I would add the new primitive to avoid breaking existing code. But I'm open to advice, particularly if we have some idea  much much the difference choices fix versus break among current Racket libraries and packages.

**Comment by michaelballantyne** [comment]
Fixed by 816c264 as discussed in #3927.
